pkg/cmd/rollbackcopy/etcdclientutils.go

```
@@ -0,0 +1,137 @@
+}
+
+func getClusterVersionAndUpgradeInfo(cli *clientv3.Client) (string, bool, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	clusterversionKey := "/kubernetes.io/config.openshift.io/clusterversions/version"
+	resp, err := cli.Get(ctx, clusterversionKey)
+	if err != nil {
+		return "", false, err
+	}
+	klog.Info("### Response for client GET: ", resp)
+	if len(resp.Kvs) != 1 {
+		klog.Errorf("Expected to get a single key from etcd, got %d", len(resp.Kvs))
+		return "", false, fmt.Errorf("Expected to get a single key from etcd, got %d", len(resp.Kvs))
+	}
+
+	klog.Info("### Value lookup: ", string(resp.Kvs[0].Value))
+	var cv map[string]interface{}
+	if err := json.Unmarshal(resp.Kvs[0].Value, &cv); err != nil {
+		return "", false, err
+	}
+
+	status := cv["status"].(map[string]interface{})
+	desired := status["desired"].(map[string]interface{})
+	klog.Info("desired version: ", desired["version"])
+
+	history := status["history"].([]interface{})
+	latestHistory := history[0].(map[string]interface{})
+	klog.Info("latest history version:", latestHistory["version"], " status: ", latestHistory["state"])
+	klog.Info("Return values: ", latestHistory["version"], " upgradeInProgress: ", desired["version"] != latestHistory["version"])
+
+	return latestHistory["version"].(string), desired["version"].(string) != latestHistory["version"].(string), nil
+}
+
+func isLeader(cli *clientv3.Client, name string) (bool, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	membersResp, err := cli.MemberList(ctx)
+	if err != nil {
+		return false, err
+	}
+	for _, member := range membersResp.Members {
+		if member.Name != name {
+			continue
+		}
+		if len(member.ClientURLs) == 0 && member.Name == "" {
+			return false, fmt.Errorf("EtcdMemberNotStarted")
+		}
+
+		resp, err := cli.Status(ctx, member.ClientURLs[0])
+		if err != nil {
+			klog.Errorf("error getting etcd member %s status: %#v", member.Name, err)
```

Wrap and return the error instead of logging